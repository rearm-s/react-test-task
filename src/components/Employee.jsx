import React, { useEffect } from 'react';
import { Button, Table } from 'react-bootstrap';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory, useParams } from 'react-router-dom';

import { Error } from './index';

import { StyledEmployee } from '../styled/Employee';
import { asyncSetEmployeeData, setError } from '../redux/ducks/employee';


const Employee = () => {

    const dispatch = useDispatch();
    const employeeData = useSelector(({employee}) => employee.employeeData);
    const history = useHistory();
    const isError = useSelector(({employee}) => employee.isError);

    const { name } =  useParams();

    useEffect(() => {
        dispatch(asyncSetEmployeeData(name));
    }, [name])

    const [position, subordinates] = employeeData;

    const handleOnBack = () => {
        history.goBack();
        dispatch(setError(false));
    }

    const handleOnSearchSubo = (name) => {
        history.push(`/employee/${name}`);
        dispatch(asyncSetEmployeeData(name));
    }

    return (
        <StyledEmployee>
            {
                isError && <Error />
            }
            <Table striped bordered hover>
                <thead>
                <tr>
                    <th>Name</th>
                    <th>Position</th>
                    <th>Subordinates</th>
                </tr>
                </thead>
                <tbody>
                <tr>
                    <td>{isError ? "-" : name}</td>
                    <td>{position ? position : "-"}</td>
                    <td>{typeof(subordinates) === "object"
                        ? subordinates["direct-subordinates"].map(subo =>
                            <ul key={subo}>
                                <li onClick={() => handleOnSearchSubo(subo)}>{subo}</li>
                            </ul>)
                        : "-"}</td>
                </tr>
                </tbody>
            </Table>
            <Button
                variant="outline-secondary"
                onClick={handleOnBack}
            >
                Go back
            </Button>
        </StyledEmployee>
    );
};

export default Employee;
