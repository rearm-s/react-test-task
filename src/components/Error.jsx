import React from 'react';
import { Alert } from 'react-bootstrap';

const Error = () => (
    <Alert variant="danger">
        Please type correct name of employee!
    </Alert>
)

export default Error;
