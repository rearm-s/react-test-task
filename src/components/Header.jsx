import React from 'react';
import { Container, Nav, Navbar } from 'react-bootstrap';
import { LinkContainer } from 'react-router-bootstrap';

const Header = () => (
        <Navbar bg="dark" variant="dark">
            <Container fluid>
                <LinkContainer to="/">
                    <Navbar.Brand>Search employee</Navbar.Brand>
                </LinkContainer>
                <Nav className="me-auto">
                    <LinkContainer to="/">
                        <Nav.Link>Home</Nav.Link>
                    </LinkContainer>
                </Nav>
            </Container>
        </Navbar>
);

export default Header;
