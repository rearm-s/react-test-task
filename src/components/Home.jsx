import React, { useEffect  } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Table } from 'react-bootstrap';
import { useHistory } from 'react-router-dom';

import { Error, SearchBox } from './index';

import { StyledHome } from '../styled/Home';
import { asyncSetData, setEmployeeName, setError } from '../redux/ducks/employee';

const Home = () => {

    const dispatch = useDispatch();
    const employeeName = useSelector(({employee}) => employee.employeeName);
    const data = useSelector(({employee}) => employee.data);
    const isError = useSelector(({employee}) => employee.isError);
    const history = useHistory();

    useEffect(() => {
        dispatch(asyncSetData());
    }, [])

    const handleOnChange = (e) => {
        dispatch(setEmployeeName(e.target.value));
    }

    const handleOnSearch = () => {
        if(!data.includes(employeeName)) {
            dispatch(setError(true));
            dispatch(setEmployeeName(''));
        } else {
            history.push(`/employee/${employeeName}`);
            dispatch(setError(false));
            dispatch(setEmployeeName(''));
        }
    }

    return (
        <StyledHome>
            <SearchBox employee={employeeName} onChangeEmployee={handleOnChange} onSearchEmployee={handleOnSearch}/>
            {isError &&
                <Error />
            }
            <Table striped bordered hover>
                <thead>
                <tr>
                    <th>Name of employee</th>
                </tr>
                </thead>
                <tbody>
                {data &&
                    data.map(emp => {
                        return <tr key={emp}>
                            <td>{emp}</td>
                        </tr>
                    })
                }
                </tbody>
            </Table>
        </StyledHome>
    );
};

export default Home;
