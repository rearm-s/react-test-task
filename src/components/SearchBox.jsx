import React  from "react";
import {Button, FormControl, InputGroup} from "react-bootstrap";

export default ({employee, onChangeEmployee, onClickButton}) => (
        <InputGroup className="mb-3">
            <FormControl
                value={employee}
                onChange={onChangeEmployee}
                placeholder="Type employee name"
            />
            <Button
                variant="outline-secondary"
                onClick={onClickButton}
            >
                Search
            </Button>
        </InputGroup>
)
