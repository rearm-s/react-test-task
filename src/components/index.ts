export { default as Employee } from './Employee';
export { default as Error } from './Error';
export { default as Header } from './Header';
export { default as Home } from './Home';
export { default as SearchBox } from './SearchBox';
