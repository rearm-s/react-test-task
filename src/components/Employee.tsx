import React, { useEffect } from 'react';
import { Button, Table } from 'react-bootstrap';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory, useParams } from 'react-router-dom';
import { Dispatch } from 'redux';

import { Error } from './index';

import { StyledEmployee } from '../styled/Employee';
import { EmployeeAction, getEmployeeData, setError } from '../redux/ducks/employee';
import { RootState } from '../redux/store';

interface RouteParams {
    name: string;
}

const Employee = (): JSX.Element => {

    const dispatch: Dispatch<EmployeeAction> = useDispatch();
    const employeeData = useSelector(({employee}: RootState) => employee.employeeData);
    const history = useHistory();
    const isError = useSelector(({employee}: RootState) => employee.isError);

    const {name} = useParams<RouteParams>();

    useEffect(() => {
        dispatch(getEmployeeData(name));
    }, [name]);

    const [position, subordinates] = employeeData;

    const handleOnBack = () => {
        history.goBack();
        dispatch(setError(false));
    };

    const handleOnSearchSubo = (name: string) => {
        history.push(`/employee/${name}`);
        dispatch(getEmployeeData(name));
    };

    return (
        <StyledEmployee>
            {isError && (
                <Error/>
            )}
            <Table striped bordered hover>
                <thead>
                <tr>
                    <th>Name</th>
                    <th>Position</th>
                    <th>Subordinates</th>
                </tr>
                </thead>
                <tbody>
                <tr>
                    <td>{isError ? "-" : name}</td>
                    <td>{position ? position : "-"}</td>
                    <td>{typeof (subordinates) === "object"
                        ? subordinates["direct-subordinates"].map((subo: string) =>
                            <ul key={subo}>
                                <li onClick={() => handleOnSearchSubo(subo)}>{subo}</li>
                            </ul>)
                        : "-"}</td>
                </tr>
                </tbody>
            </Table>
            <Button
                variant="outline-secondary"
                onClick={handleOnBack}
            >
                Go back
            </Button>
        </StyledEmployee>
    );
};

export default Employee;
