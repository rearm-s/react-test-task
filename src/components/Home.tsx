import React, { useEffect  } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Table } from 'react-bootstrap';
import { useHistory } from 'react-router-dom';
import { Dispatch } from "redux";

import { Error, SearchBox } from './index';

import { StyledHome } from '../styled/Home';
import { EmployeeAction, getData, setEmployeeName, setError } from '../redux/ducks/employee';
import { RootState } from '../redux/store';

const Home = (): JSX.Element => {

    const dispatch: Dispatch<EmployeeAction> = useDispatch();
    const employeeName = useSelector(({employee}: RootState) => employee.employeeName);
    const data = useSelector(({employee}: RootState) => employee.data);
    const isError = useSelector(({employee}: RootState) => employee.isError);
    const history = useHistory();

    useEffect(() => {
        dispatch(getData());
    }, [])

    const handleOnChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        dispatch(setEmployeeName(e.target.value));
    }

    const handleOnSearch = () => {
        if(!data.includes(employeeName)) {
            dispatch(setError(true));
            dispatch(setEmployeeName(''));
        } else {
            history.push(`/employee/${employeeName}`);
            dispatch(setError(false));
            dispatch(setEmployeeName(''));
        }
    }

    return (
        <StyledHome>
            <SearchBox employee={employeeName} onChangeEmployee={handleOnChange} onSearchEmployee={handleOnSearch}/>
            {isError &&
            <Error />
            }
            <Table striped bordered hover>
                <thead>
                <tr>
                    <th>Name of employee</th>
                </tr>
                </thead>
                <tbody>
                {data &&
                data.map((emp: any) => {
                    return <tr key={emp}>
                        <td>{emp}</td>
                    </tr>
                })
                }
                </tbody>
            </Table>
        </StyledHome>
    );
};

export default Home;
