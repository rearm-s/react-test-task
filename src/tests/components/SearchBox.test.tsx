import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { SearchBox } from './../../components';

configure({adapter: new Adapter()});

const employee = 'test';

const handleOnChange = jest.fn();

const handleOnSearch = jest.fn();

let wrapper: any;

describe('SearchBox', () => {

    beforeAll(() => {
        wrapper = shallow(<SearchBox
            employee={employee}
            onChangeEmployee={() => handleOnChange('testAfterHandler')}
            onSearchEmployee={handleOnSearch}/>);
    });

    it('should take a snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

});
