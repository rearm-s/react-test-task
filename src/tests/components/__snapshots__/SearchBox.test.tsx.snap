// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox should take a snapshot 1`] = `ShallowWrapper {}`;
